### ISSUE TITLE
USER_HZ counting incorrectly on Linux

### ISSUE THREAD

**Issue by risik** [oshi/oshi#683]
USER_HZ constant calculated incorrectly and so getKernelTime, getUserTime and getUpTime methods of OSProcess class returns incorrect results.

Code example:

    private void run() {
        for (int i = 0; i < 100000000; i++) {
            System.out.println("-----------------------------------------");
            printProcessCpuInfo();
            System.out.println("=========================================");
        }
    }

    void printProcessCpuInfo() {
        final SystemInfo si = new SystemInfo();
        final OperatingSystem os = si.getOperatingSystem();

        // this process
        final int processId = os.getProcessId();
        final OSProcess osProcess = os.getProcess(processId);

        try {
            Thread.sleep(1000);
        }
        catch (InterruptedException ignored) {
        }

        System.out.println("now: " + new Date().getTime());
        System.out.println("System.nanoTime(): " + System.nanoTime());
        System.out.println("process.start_time: " + osProcess.getStartTime()); // here OK
        System.out.println("process.cpu.kernel_usage_ms: " + osProcess.getKernelTime()); // it's too large
        System.out.println("process.cpu.user_usage_ms: " + osProcess.getUserTime());  // it's too large
        System.out.println("process.cpu.up_time: " + osProcess.getUpTime()); // here always 1
        System.out.println("process.cpu.usage_percent: " + osProcess.calculateCpuPercent()); // here we have very large values due getUpTime works incorrectly
    }

Output example:

----------------------------------------
now: 1542464204617
System.nanoTime(): 90196842438436
process.start_time: 1542464203607
process.cpu.kernel_usage_ms: 1705
process.cpu.user_usage_ms: 10029
process.cpu.up_time: 1
process.cpu.usage_percent: 11734.0
=========================================
-----------------------------------------
now: 1542464205623
System.nanoTime(): 90197848765779
process.start_time: 1542464204618
process.cpu.kernel_usage_ms: 1735
process.cpu.user_usage_ms: 10058
process.cpu.up_time: 1
process.cpu.usage_percent: 11793.0
=========================================

In this time HTOP displays time 4.01 seconds.

before stop application:
$ cat /proc/14826/stat | awk '{print "PPID: " $4 " USER_TIME: " $14 " KERNEL_TIME: " $15 " PRIORITY: " $18 " THREAD_COUNT: " $20 " START_TIME: " $22 " VSZ: " $23 " RSS: " $24}'
PPID: 14754 USER_TIME: 290 KERNEL_TIME: 46 PRIORITY: 20 THREAD_COUNT: 19 START_TIME: 26330527 VSZ: 11364433920 RSS: 27482

$ cat /proc/uptime 
263616.17 276045.52

uptime in htop: 3 days + 1hour...

While debugging library code I see that USER_HZ=30 but according my calculation it must be approx 10.

My platform:

$ uname -orp
4.15.0-39-generic x86_64 GNU/Linux

$ java -version
java version "1.8.0_191"
Java(TM) SE Runtime Environment (build 1.8.0_191-b12)
Java HotSpot(TM) 64-Bit Server VM (build 25.191-b12, mixed mode)

**Comment by dbwiddis** [comment]
Thanks for the detailed report and debugging info.
OSHI calculates the HZ value by the ratio of the system idle time in seconds (second field of `/proc/uptime`) and system idle time (fourth numeric field in first ("cpu") line of `/proc/stat`).

Can you please post the `/proc/uptime` (again) and `/proc/stat` outputs so I can review these calculations?

**Comment by dbwiddis** [comment]
The second output looks as expected.  HZ calculates to 100 (605610795 / 6049845.5) and the total CPU usage of 2.0 is correct for a 2 CPU system.

The first output, though, is odd, showing 7454384 idle ticks in 302087.94 seconds, for a HZ of 24.676 (rounds to 24, which you saw in your breakpoint).

Looking at `now` of 1542539751041 ms vs. `btime` of 1542200609 secs we get an up time of about 339142 seconds, which generally matches `/proc/uptime` (accounting for a ~16 min time difference between commands).  

Total ticks for all 4 processors is 17429307.  Matching that to the uptime of 338162.59 seconds yields a HZ of 51.54 ticks per second for all 4 CPUs, or 12.89 per CPU.

Bottom line:  the tick values in `/proc/stat` do not make sense for a constant USER_HZ value.  Was this value changed after bootup?

**Comment by dbwiddis** [comment]
Also of interest, the different CPUs have accrued different tick totals.  On the second (correct) CentOS result, cpu0 and cpu1 values are within 0.03% of each other. But for the first result, cpu0 has accrued 9944556 ticks, cpu1 2566867, cpu2 2530404 and cpu3 2387470.  It seems all 4 cpus have not been running the whole time.  Is this some hardware that can dynamically change the number of CPUs running?  A VM that has had more CPUs assigned recently?

**Comment by risik** [comment]
1. After reboot my laptop:

app output:
now: 1542613228510
System.nanoTime(): 447169262781
process.start_time: 1542613042893
process.cpu.kernel_usage_ms: 438
process.cpu.user_usage_ms: 2632
process.cpu.up_time: 184611
process.cpu.usage_percent: 0.01662956161875511

$  cat /proc/uptime
459.15 1212.36

$ cat /proc/stat
cpu  51175 320 9534 120817 1746 0 386 0 0 0
cpu0 13042 24 2649 29834 423 0 61 0 0 0
cpu1 12189 93 2387 30795 424 0 146 0 0 0
cpu2 13136 70 2226 29965 424 0 125 0 0 0
cpu3 12806 131 2270 30222 473 0 52 0 0 0
intr 1510763 11 1426 0 0 0 0 0 0 1 1228 0 0 56059 0 0 0 0 0 0 13 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 6430 48079 77197 13 105155 69585 62720 75555 47 1322 917 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
ctxt 3259837
btime 1542612781
processes 7980
procs_running 1
procs_blocked 0
softirq 720487 119855 197117 107 48582 252 0 1529 126260 0 226785

on moment of this output htop dispays 3.01 seconds eated by this process. It the same to  2632+438=3070 that I got from the application.

2.
> Also of interest, the different CPUs have accrued different tick totals. On the second (correct) CentOS result, cpu0 and cpu1 values are within 0.03% of each other. But for the first result, cpu0 has accrued 9944556 ticks, cpu1 2566867, cpu2 2530404 and cpu3 2387470. It seems all 4 cpus have not been running the whole time. Is this some hardware that can dynamically change the number of CPUs running? A VM that has had more CPUs assigned recently?

This is not an VM but laptop. It sometimes going to sleep and control CPU's frequency (eg decrease frequency due throttling)... not sure may be this matter.

3.
To test I manually turn off 3 core of my CPU in this way:
echo 0 | sudo tee /sys/devices/system/cpu/cpu1/online
and run again after ~25 minutes:

app output:
now: 1542615662897
System.nanoTime(): 2881555611292
process.start_time: 1542615346775
process.cpu.kernel_usage_ms: 510
process.cpu.user_usage_ms: 2500
process.cpu.up_time: 315104
process.cpu.usage_percent: 0.00955240174672489

htop time 3.01 seconds.

$ cat /proc/uptime
2929.51 4571.38

$ cat /proc/stat
cpu  123057 344 32970 452469 2317 0 1504 0 0 0
cpu0 65600 46 20758 200441 1091 0 876 0 0 0
intr 4408277 11 8152 0 0 0 0 0 0 1 2194 0 0 239467 0 0 0 0 0 0 13 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 12566 85393 90913 13 332616 71031 63804 76627 47 4789 925 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
ctxt 11073471
btime 1542612781
processes 9614
procs_running 1
procs_blocked 0
softirq 2399179 402084 517120 580 87580 1480 3 7341 237244 0 1145747

4. Just after test 3 I turned on all cores and after more 1h30m I repeated this test

Again I stop application then htop displays 3.00 seconds that app eat.

App output:
now: 1542620994207
System.nanoTime(): 8212866321569
process.start_time: 1542620993201
process.cpu.kernel_usage_ms: 500
process.cpu.user_usage_ms: 2988
process.cpu.up_time: 1
process.cpu.usage_percent: 3488.0

* process.cpu.kernel_usage_ms + process.cpu.user_usage_ms = 3.5 seconds it's again too large comparing to htop. 
* process.cpu.up_time again always 1.

$ cat /proc/uptime
8244.89 21568.89

$ cat /proc/stat
cpu  394056 808 104495 1871032 9535 0 5067 0 0 0
cpu0 136990 150 40075 632573 1956 0 1783 0 0 0
cpu1 80369 199 21048 419268 752 0 904 0 0 0
cpu2 83064 191 21090 416718 3209 0 965 0 0 0
cpu3 93632 267 22281 402472 3616 0 1414 0 0 0
intr 14307120 11 32339 0 0 0 0 0 0 1 4223 0 0 943621 0 0 0 0 0 0 13 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 25685 165080 102234 13 1466597 81738 74544 87970 47 18685 931 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
ctxt 33390414
btime 1542612781
processes 13758
procs_running 1
procs_blocked 0
softirq 7686653 1856934 1755858 1575 170982 4080 3 26108 1282174 0 2588939

**Comment by dbwiddis** [comment]
Thanks, looks like we've figured out the issue.   OSHI's HZ calculation uses the total number of cores assuming it's been a constant, but that assumption isn't always valid. 

Originally we looked at every process to identify the newest one and used that for the HZ calculation, but that had a small performance hit.  Then we guessed at the highest PID process but that proved unreliable.  Using total idle time should have been a nice solution but clearly misses this use case.  

Ultimately I guess we need to make Hz a variable and update it more than once if the initial guess proves wrong.

**Comment by dbwiddis** [comment]
So after all this, I find out that `getconf CLK_TCK` returns USER_HZ from the command line... go figure!
